src/compiler/utils/string_with_sourcemap.ts

```diff
@@ -288,3 +290,38 @@ export function apply_preprocessor_sourcemap(filename: string, svelte_map: Sourc
 
 	return result_map as SourceMap;
 }
+
+// parse attached sourcemap in processed.code
+export function parse_attached_sourcemap(processed: Processed, tag_name: 'script' | 'style'): void {
+	const r_in = '[#@]\\s*sourceMappingURL\\s*=\\s*(\\S*)';
+	const regex = (tag_name == 'script')
+		? new RegExp('(?://'+r_in+')|(?:/\\*'+r_in+'\\s*\\*/)$')
+		: new RegExp('/\\*'+r_in+'\\s*\\*/$');
+	function throw_error(message) {
+		throw 'error: ' + message + '\nprocessed.code:\n' +
+			(processed.code.length < 100 ? processed.code : processed.code.slice(0, 100)+' [...]'); // help to find preprocessor
+	}
+	processed.code = processed.code.replace(regex, (_, match1, match2) => {
+		const map_url = (tag_name == 'script') ? (match1 || match2) : match1;
+		const map_data = (map_url.match(/data:(?:application|text)\/json;(?:charset[:=]\S+?;)?base64,(\S*)/) || [])[1];
+		if (map_data) {
+			// sourceMappingURL is data URL
+			if (processed.map) {
+				throw_error('not implemented. '+
+					'found sourcemap in both processed.code and processed.map. '+
+					'please pass only one sourcemap.');
+			}
+			processed.map = b64dec(map_data); // use attached sourcemap
+			return ''; // remove from processed.code
+		}
+		// sourceMappingURL is file path or URL
+		if (!processed.map) {
+			// TODO show warning? silently ignore?
+			throw_error('value error. processed.map is empty, '+
```

Comment:
if `processed.code` is pointing to a file with a sourcemap wouldn't we just use that? why do we also need a sourcemap in `processed.map`?

Reply:
that file does not exist, usually
also svelte should not access the file system

having `//# sourceMappingURL=Component.svelte.js.map` attached to `processed.code`
but no sourcemap in `processed.map` may indicate an error

Reply:
Perhaps we should change this to say something like "Sourcemap expected in processed.map, but found in processed.code. Please update your preprocessor". Would that make sense to you?

Is it possible we can include which preprocessor it is? I think it may be hard to respond to without knowing that

Reply:
> Sourcemap ... found in processed.code

is wrong, cos there is only the sourcemap path `Component.svelte.js.map` but no sourcemap data

other suggestions than ... ?

> value error. processed.map is empty, but processed.code has attached sourcemap path "Component.svelte.js.map". please make your preprocessor return a sourcemap.

> Is it possible we can include which preprocessor it is?

no, currently we do not track preprocessors as i suggested in

https://github.com/sveltejs/svelte/blob/dbd184c38c0f517107316835db948619ff4cf8c0/src/compiler/preprocess/index.ts#L228-L230

but usually this is not a problem for end-users, but for plugin developers

to make easier the debugging of preprocessors, we could add a debug mode (or trace mode)
where all code + sourcemaps are pretty-printed to console
(thats what i do manually in that case)

Reply:
How about `'Sourcemap expected in processed.map, but found path ' + path + ' in processed.code. Please update your preprocessor to use processed.map instead'`

Reply:
> Please update your preprocessor to use processed.map instead

is not 100% correct, since we tolerate the combination of sourcemap data in `processed.map` and sourcemap path in `processed.code`. use case: typescript preprocessor, where we cannot disable the sourcemap-path-attachment

good combinations:

* sourcemap data in processed.map
* sourcemap data in processed.code (as base64 data URL)
* sourcemap data in processed.map + sourcemap path in processed.code (path is ignored since we cannot access the file system)

bad combinations:

* sourcemap data in processed.map + sourcemap data in processed.code (ambiguous if sourcemaps are not equal, we dont know in what time-order we should combine two different sourcemaps)
* sourcemap path in processed.code (no sourcemap data anywhere. if the code was not modified, then there should be no sourcemap in any form, also no sourcemap path)

how about ... ?

> Found sourcemap path "Component.svelte.js.map" in processed.code, but no sourcemap data. Please update your preprocessor to return a sourcemap.

or generalize the warning: if code was changed, but no sourcemap was returned
sure, sourcemaps are optional, but missing sourcemaps are a pain for debugging

Reply:
Ok. That message sounds good to me. I might tweak it just a bit to be:

>Found sourcemap path "Component.svelte.js.map" in processed.code, but no sourcemap data. Please update your preprocessor to return sourcemap data directly.